controllers/metric_controller.go

```diff
@@ -208,49 +207,145 @@ func (r *MetricReconciler) collectMetrics(ctx context.Context, t *redskyv1beta1.
 
-		if m.URL != "" {
-			return target, nil
+		// Metric errors contain additional information which should be logged for debugging
+		if merr, ok := err.(*metric.CaptureError); ok {
+			log.Error(merr, "Metric collection failed", "address", merr.Address, "query", merr.Query, "completionTime", merr.CompletionTime)
 		}
+	} else if v.AttemptsRemaining > 0 {
+		// Decrement the attempts remaining
+		v.AttemptsRemaining--
+	}
 
-		// Default label selector to `app: prometheus` if not specified
-		labelSelector := m.Selector
-		if labelSelector == nil && m.Type == redskyv1beta1.MetricPrometheus {
-			labelSelector = &metav1.LabelSelector{
-				MatchLabels: map[string]string{"app": "prometheus"},
-			}
-		}
+	// Record the update
+	return controller.RequeueConflict(r.Update(ctx, t))
+}
 
-		sel, err := meta.MatchingSelector(labelSelector)
-		if err != nil {
-			return nil, err
-		}
+// target looks up the Kubernetes object (if any) associated with a metric.
+func (r *MetricReconciler) target(ctx context.Context, t *redskyv1beta1.Trial, m *redskyv1beta1.Metric) (runtime.Object, error) {
+	if m.Type != redskyv1beta1.MetricKubernetes && m.Type != "" {
+		return nil, nil
+	}
+
+	// Get the target reference, default to the trial itself if there is no reference
+	targetRef := m.Target
+	if m.Target == nil || m.Target.Kind == "" {
+		return t, nil
+	}
 
-		if err = r.List(ctx, target, client.InNamespace(namespace), sel); err != nil {
+	// If no explicit namespace is specified, use the trial namespace
+	if targetRef.Namespace == "" {
+		targetRef.Namespace = t.Namespace
+	}
+
+	// If a name is specified, just get a single object
+	if targetRef.Name != "" {
+		target := &unstructured.Unstructured{}
+		target.SetGroupVersionKind(targetRef.GroupVersionKind())
+		if err := r.Get(ctx, targetRef.NamespacedName(), target); err != nil {
 			return nil, err
 		}
-
 		return target, nil
-	default:
-		// Assume no target is necessary
-		return nil, nil
 	}
+
+	// Convert the selector from a Kubernetes object to something the client can use
+	sel, err := meta.MatchingSelector(targetRef.LabelSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fetch the list of matching resources
+	target := &unstructured.UnstructuredList{}
+	target.SetGroupVersionKind(targetRef.GroupVersionKind())
+	if err := r.List(ctx, target, client.InNamespace(targetRef.Namespace), sel); err != nil {
+		return nil, err
+	}
+	return target, nil
+}
+
+// applyMetricDefaults fills in default values for the supplied metric.
+func (r *MetricReconciler) applyMetricDefaults(ctx context.Context, t *redskyv1beta1.Trial, m *redskyv1beta1.Metric) error {
+	// Give Prometheus metrics a default URL
+	if m.Type == redskyv1beta1.MetricPrometheus && m.URL == "" {
+		m.URL = fmt.Sprintf("http://redsky-%s-prometheus:9090/", t.Namespace)
+	}
+
+	// Default to the trial namespace
+	if m.Target != nil && m.Target.Namespace == "" {
+		// TODO How do we exclude non-namespaced kinds from this default?
```

Comment:
should we drop this TODO

testing via kubectl, `kubectl get clusterole` vs `kubectl get clusterrole -n kube-system` both seem to work